mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,188 @@
+    """
+    list_metric = []
+    for res in results:
+        list_metric.append(res[2][metric]['values'][-1])
+
+    list_metric = np.array(list_metric)
+
+    # get best runs
+    if metric_direction:
+        best_metric = list_metric.max()
+        best_idx = list_metric.argmax()
+    else:
+        best_metric = list_metric.min()
+        best_idx = list_metric.argmin()
+
+    return best_metric, best_idx
+
+
+def launch_grid(linear_kwargs, regularizer_kwargs, reconstructor_kwargs,
+                optimizer_kwargs, compare_metric_details=None, n_jobs=1,
+                verbose=0, **kwargs):
+    """This function launches off reconstruction for a grid specified
+    through use of kwarg dictionaries.
+
+    Dictionary Convention
+    ---------------------
+    These dictionaries each defined to follow the convention:
+    Each dictionary has a key `init_class` that specifies the
+    initialization class for the operator (exception to
+    this is 'optimizer_kwargs'). Later we have key `args` that holds
+    all the input arguments that can be passed as a keyword dictionary.
+    Each value in this keyword dictionary ,ust be a list of all
+    values you want to search in gridsearch.
+
+    This function finds the search space of parameters and
+    sets up right parameters for '_reconstruct_case' function.
+    Please check the example code for more details.
+
+    Parameters
+    ----------
+    linear_kwargs: dict
+        dictionary for linear operator parameters
+    regularizer_kwargs: dict
+        dictionary for regularizer operator parameters
+    reconstructor_kwargs: dict
+        dictionary for reconstructor operator parameters
+    optimizer_kwargs: dict
+        dictionary for optimizer key word arguments
+    compare_metric_details: dict default None
+        dictionary that holds the metric to be compared and metric
+        direction please refer to `gather_result` documentation.
+    n_jobs: int, default 1
+        number of parallel jobs for execution
+    verbose: int default 0
+        Verbosity level
+        0 => No debug prints
+        1 => View best results if present
+
+    **kwargs: keyword arguments that are passes to reconstruction
+        these holds extra arguments that are passed to
+        '_reconstruct_case' function.
+        important arguments are 'ksapce_data' and 'fourier_kwargs'
+    """
+    # Convert non-list elements to list so that we can create
+    # search space
+    all_reformatted_kwargs = []
+    for specific_kwargs in [linear_kwargs, regularizer_kwargs,
+                            reconstructor_kwargs, optimizer_kwargs]:
+        for key, value in specific_kwargs['args'].items():
+            if not isinstance(value, list) and \
+                    not isinstance(value, np.ndarray):
+                specific_kwargs['args'][key] = [value]
+        all_reformatted_kwargs.append(specific_kwargs)
+    # Create Search space
```

Comment:
Reformatted most codes here as we already have loop above, moved most helper variable creations to there

Reply:
I don't understand what you wrote.

Reply:
I mean to say that I got rid of that statement, and used earlier loop in code

Reply:
yes but there are still some instance below

Reply:
Hmm, are you refering to L231? I dont know how to get rid of it there, the length is number of test cases, at best I can split it to:
```python
num_test_cases = len(cross_product_list)
for i in range(num_test_cases)
    ....
```

Reply:
Well in this for example:

```python
for i in range(number_of_test_cases):
    # Iterator for i-th test case
    iterator = iter(cross_product_list[i])
```

First the comment is a bit redundant. Second why not do:
```python
for test_case in cross_product_list:
    iterator = iter(test_case)
```

since you don't use `i` afterwards (and even if you were you could always use `zip` or `enumerate` in last resort).

Same goes for the parallel execution.